From 2ceb46d976ae1f971dd70270072fc17e7524e93d Mon Sep 17 00:00:00 2001
From: Oliver Heger <oheger@apache.org>
Date: Sun, 23 Sep 2007 16:20:24 +0000
Subject: [PATCH] CONFIGURATION-296: The root node of a XMLConfiguration is now
 properly initialized with a reference to the XML document's root element

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/configuration/trunk@578579 13f79535-47bb-0310-9956-ffa450edef68
---
 .../configuration/XMLConfiguration.java        |  7 ++++++-
 .../configuration/TestXMLConfiguration.java    | 18 ++++++++++++++++++
 xdocs/changes.xml                              |  4 ++++
 3 files changed, 28 insertions(+), 1 deletion(-)

diff --git a/src/java/org/apache/commons/configuration/XMLConfiguration.java b/src/java/org/apache/commons/configuration/XMLConfiguration.java
index c162a2b9e0..2999768f08 100644
--- a/src/java/org/apache/commons/configuration/XMLConfiguration.java
+++ b/src/java/org/apache/commons/configuration/XMLConfiguration.java
@@ -419,7 +419,12 @@ public void initProperties(Document document, boolean elemRefs)
             setPublicID(document.getDoctype().getPublicId());
             setSystemID(document.getDoctype().getSystemId());
         }
+
         constructHierarchy(getRoot(), document.getDocumentElement(), elemRefs);
+        if (elemRefs)
+        {
+            getRoot().setReference(document.getDocumentElement());
+        }
     }
 
     /**
@@ -1242,7 +1247,7 @@ private static void updateAttribute(Node node, Element elem, String name, char l
          *
          * @param node the affected node
          * @param name the name of the attribute
-         * @param listDelimiter the delimiter vor attributes with multiple values
+         * @param listDelimiter the delimiter for attributes with multiple values
          */
         static void updateAttribute(Node node, String name, char listDelimiter)
         {
diff --git a/src/test/org/apache/commons/configuration/TestXMLConfiguration.java b/src/test/org/apache/commons/configuration/TestXMLConfiguration.java
index c7d30c974e..01b2088f14 100644
--- a/src/test/org/apache/commons/configuration/TestXMLConfiguration.java
+++ b/src/test/org/apache/commons/configuration/TestXMLConfiguration.java
@@ -301,6 +301,24 @@ public void testSetRootAttribute() throws ConfigurationException
         checkSavedConfig(checkConf);
         assertTrue("Attribute not found after save", checkConf
                 .containsKey("[@test]"));
+        checkConf.setProperty("[@test]", "newValue");
+        checkConf.save();
+        conf = checkConf;
+        checkConf = new XMLConfiguration();
+        checkConf.setFile(testSaveConf);
+        checkSavedConfig(checkConf);
+        assertEquals("Attribute not modified after save", "newValue", checkConf
+                .getString("[@test]"));
+    }
+
+    /**
+     * Tests whether the configuration's root node is initialized with a
+     * reference to the corresponding XML element.
+     */
+    public void testGetRootReference()
+    {
+        assertNotNull("Root node has no reference", conf.getRootNode()
+                .getReference());
     }
 
     public void testAddList()
diff --git a/xdocs/changes.xml b/xdocs/changes.xml
index ca6b55ea67..355e3a4cc3 100644
--- a/xdocs/changes.xml
+++ b/xdocs/changes.xml
@@ -23,6 +23,10 @@
 
   <body>
     <release version="1.5-SNAPSHOT" date="in SVN" description="">
+      <action dev="oheger" type="fix" issue="CONFIGURATION-296">
+        A bug in XMLConfiguration caused that attributes of the root element
+        could not be changed. This has been fixed.
+      </action>
       <action dev="oheger" type="add" issue="CONFIGURATION-290">
         A new method registerEntityId() was added to XMLConfiguration, which
         allows to register URLs for entities. A new default implementation of
